### PR TITLE
[FIRRTL] Add canonicalizers for verification ops

### PR DESCRIPTION
**Pull request by youngar**
This replicates some assertions implemented in the SV dialect.  We need
to move these into the FIRRTL dialect so that they may be removed before
`IMDeadCodeElim` runs.

### CODE REVIEW

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 1]
```diff
@@ -1833,3 +1833,52 @@ LogicalResult RegOp::canonicalize(RegOp op, PatternRewriter &rewriter) {
...
+
+//===----------------------------------------------------------------------===//
+// Verification Ops.
+//===----------------------------------------------------------------------===//
+
+static LogicalResult eraseIfZeroOrNotZero(Operation *op, Value predicate,
+                                          Value enable,
+                                          PatternRewriter &rewriter,
+                                          bool eraseIfZero) {
+  // If the assertion is never enabled, delete it.
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1845]: Supreme nit: `s/assertion/verification op/`.

**Review by seldridge — APPROVED** [review]
LGTM

File: test/Dialect/FIRRTL/canonicalization.mlir [hunk 2]
```diff
@@ -2385,4 +2385,19 @@ firrtl.module @AnnotationsBlockRemoval(
   firrtl.strictconnect %b, %w : !firrtl.uint<1>
 }
 
+// CHECK-LABEL: firrtl.module private @Verification({{.+}}) {
+// CHECK-NEXT:  }
```
[the remark this answers is not in the document]
Comment (youngar) [reply test/Dialect/FIRRTL/canonicalization.mlir]: Not sure exactly what this meant, updated it to check that each statement was removed, but did you want me to put this all in the top part of the check?

[the remark this answers is not in the document]
Comment (seldridge) [reply test/Dialect/FIRRTL/canonicalization.mlir]: I was thinking more like putting three `CHECK-NOT`s at the top after the label. Alternatively, the original is fine with a comment that it's checking for removal.

The explicit check nots are fine, but the arguments aren't necessary.

File: lib/Dialect/FIRRTL/FIRRTLFolds.cpp [hunk 3]
```diff
@@ -1833,3 +1833,52 @@ LogicalResult RegOp::canonicalize(RegOp op, PatternRewriter &rewriter) {
...
+
+void AssumeOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                           MLIRContext *context) {
+  results.add(canonicalizeImmediateVerifOp<AssumeOp>);
+}
+
+void CoverOp::getCanonicalizationPatterns(RewritePatternSet &results,
+                                          MLIRContext *context) {
+  results.add(canonicalizeImmediateVerifOp<CoverOp, /* EraseIfZero = */ true>);
+}
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/FIRRTLFolds.cpp:1884]: super nit: (I guess this is actually from my diff though :))
```suggestion
}

```

**Review by uenoku — APPROVED** [review]
LGTM, thank you for finishing this!

**Pull request merged**